src/DataSet.cpp

```diff
@@ -263,3 +273,129 @@ void DataSet::registerPsm(PSMDescription* myPsm) {
+    if (!isTab) {
+      std::cerr << file_name << " is not comma delimited!\n" << std::endl;
+    }
+    return isTab;
+}
+
+bool TabFileValidator::isTabFiles(std::vector<std::string> files) {
+  for(const string &file : files) {
+      if(!isTabFile(file)) {
+        return false;
+      } 
+    };
+    return true;
+}
+
+std::string TabFileValidator::getDecoyPrefix(std::vector<std::string> fileList) {
+  std::string decoy_prefix;
+  for(const string &file : fileList) {
+        decoy_prefix = detect_decoy_prefix(file);
+        break;
+      };
+      return decoy_prefix;
+}
+
+void TabFileValidator::getProteinIndex(std::string file_name,int* proteinIndex,int* labelIndex) {
+  // open C++ stream to file
+    std::ifstream file(file_name.c_str());
+
+    // file not opened, return false
+    if(!file.is_open()) {
+      return ;
+    }
+
+    int columnIndex = 0;
+    std::string hederRow;
+    getline(file, hederRow);
+    TabReader reader(hederRow);
+      while (!reader.error()) {
+        std::string optionalHeader = reader.readString();
+        if (optionalHeader.find("Proteins") != std::string::npos) { 
+          *proteinIndex = columnIndex;
+        } else if (optionalHeader.find("Label") != std::string::npos){
+          *labelIndex = columnIndex;
+        }
+        columnIndex++;
+     }
+}
+
+std::string TabFileValidator::findDecoyPrefix(std::string file_name, int proteinIndex, int labelIndex) {
+  // open C++ stream to file
+    std::ifstream file(file_name.c_str());
+    // file not opened, return false
+    if(!file.is_open()) return "error";
+    // read a line from the file  
+
+    /* Skip header */
+    std::string nextRow;
+    getline(file, nextRow);
+      
+    while(getline(file, nextRow)) {
+
+        TabReader readerRow(nextRow);
+        int col = 0;
+        bool isDecoy = false;
+
+        while (!readerRow.error()) {
+          std::string value = readerRow.readString();
+          if (col == labelIndex && value == "-1") {
+            isDecoy = true;
+          }
+          if (col == proteinIndex && isDecoy) {
+            std::string token = value.substr(0, value.find("_") + 1);
+            return token;
+          }
+          col++;
+        }
+    }
+    std::cerr << "Couldn't find protein decoy-preix, i.e., 'decoyPattern_' in protein ids" << std::endl;
+    return "error";
+  
+}
+
+std::string TabFileValidator::detect_decoy_prefix(std::string file_name) {
+
+    int proteinIndex=-1;
+    int labelIndex=-1;
+  
+    getProteinIndex(file_name, &proteinIndex, &labelIndex);
+    if (proteinIndex==-1 || labelIndex==-1) {
+      std::cerr << "Couldn't find Protein header in tab-file" << std::endl;
+      return "error";
+    }
+
+    return findDecoyPrefix(file_name, proteinIndex, labelIndex);
+}
+
+bool TabFileValidator::validateTabFiles(std::vector<std::string> files, std::string* decoy_prefix) {
+
+    std::string _decoy_prefix = getDecoyPrefix(files);
```

Comment:
Don't use variable names starting with an underscore, they're reserved for other things in C++.

Reply:
fixed